Add client support for nack

quixoten opened the pull request:
I was initially against adding a nack, but I think there are some benefits to it. Using a nack allows the client to backoff more intelligently and with smaller intervals. This PR adds support to the client, but does not add support to the server. Once all clients have been upgraded to this version, support for the server side can be added.

I'm wondering if it might make sense to remove ACK completely and just rely on NACKs and a single timeout. Doing so would remove remove the possibility for a request to be duplicated at the rpc layer. It would also remove the need to receive two responses for every request. On the other hand, it would no longer cover the case that all servers stop responding (if they're all OOM killed in close proximity for example). The ACK allows us to fail in 15 seconds rather than 60 seconds in that scenario.

---

RFC @abrandoned @film42

abrandoned commented on lib/protobuf/nats/client.rb:
```diff
@@ -65,6 +71,12 @@ def send_request
         # Nats response timeout.
         retry if (retries -= 1) > 0
         raise
+      rescue NackError
```
as I was thinking last night I believe the "ideal" would be to make this throw/catch instead of rescue based; we know that we are in a "fail/retry" state and don't need to rely on an exception; also in Jruby the generation of an exception is slow/heavy

That's right ... 100x difference with a cold throw/catch from a warm rescue

```ruby
jruby-9.1.7.0 :002 > def res
jruby-9.1.7.0 :003?>   fail "derp"
jruby-9.1.7.0 :004?>   rescue => e
jruby-9.1.7.0 :005?>   nil
jruby-9.1.7.0 :006?>   end
 => :res 
jruby-9.1.7.0 :007 > def tc
jruby-9.1.7.0 :008?>   catch :something do
jruby-9.1.7.0 :009 >       end
jruby-9.1.7.0 :010?>   end
 => :tc 
jruby-9.1.7.0 :011 > def tc
jruby-9.1.7.0 :012?>   caught = false
jruby-9.1.7.0 :013?>   catch :something do
jruby-9.1.7.0 :014 >       return if caught
jruby-9.1.7.0 :015?>     caught = true
jruby-9.1.7.0 :016?>     throw :something
jruby-9.1.7.0 :017?>     end
jruby-9.1.7.0 :018?>   end
 => :tc 
jruby-9.1.7.0 :019 > Benchmark.measure do
jruby-9.1.7.0 :020 >     100_000.times do
jruby-9.1.7.0 :021 >       res
jruby-9.1.7.0 :022?>     end
jruby-9.1.7.0 :023?>   end
 => #<Benchmark::Tms:0x383bfa16 @stime=0.04999999999999999, @label="", @cstime=0.0, @real=13.571504258085042, @total=14.210000000000003, @cutime=0.0, @utime=14.160000000000002> 
jruby-9.1.7.0 :024 > Benchmark.measure do
jruby-9.1.7.0 :025 >     100_000.times do
jruby-9.1.7.0 :026 >       res
jruby-9.1.7.0 :027?>     end
jruby-9.1.7.0 :028?>   end
 => #<Benchmark::Tms:0x4229bb3f @stime=0.010000000000000009, @label="", @cstime=0.0, @real=13.664115638006479, @total=13.870000000000003, @cutime=0.0, @utime=13.860000000000003> 
jruby-9.1.7.0 :029 > Benchmark.measure do
jruby-9.1.7.0 :030 >     100_000.times do
jruby-9.1.7.0 :031 >       tc
jruby-9.1.7.0 :032?>     end
jruby-9.1.7.0 :033?>   end
 => #<Benchmark::Tms:0x543588e6 @stime=0.0, @label="", @cstime=0.0, @real=0.13869739905931056, @total=0.35999999999999943, @cutime=0.0, @utime=0.35999999999999943> 
```

quixoten replied:
Would it be worth avoiding all together and just returning one of the symbols `:ack_timeout`, `:nack`, `:resp_timeout` when those conditions occur?

abrandoned replied:
even in MRI it's 4x faster to throw/catch (just FYI)

quixoten commented:
I've updated the implementation to remove the fail/rescue from ack timeout and nack handling. I also modified the behavior slightly so that we return immediately if the response is received before the ack. I don't think we care what happens to the ack when the response is delivered first.

film42 commented on lib/protobuf/nats/client.rb:
```diff
@@ -5,6 +5,11 @@
 module Protobuf
   module Nats
     class Client < ::Protobuf::Rpc::Connectors::Base
+      NACK_BACKOFF_INTERVALS = [0, 1, 3, 5, 10]
```
I would really like this to be configurable. This seems fine for now though.

film42 commented on lib/protobuf/nats/client.rb:
```diff
@@ -100,9 +118,12 @@ def nats_request_with_two_responses(subject, data, opts)
-          fail ::NATS::IO::Timeout if first_message.nil?
+          return :ack_timeout if first_message.nil?
+          return :nack if first_message.data == ::Protobuf::Nats::Messages::NACK
+          return first_message.data unless first_message.data == ::Protobuf::Nats::Messages::ACK
```
This should break a test. We need to have an ACK and a response message in any order and we check for that below.

EDIT: I see you changed behavior in the tests... don't think this is a good idea.

abrandoned commented on lib/protobuf/nats/client.rb:
```diff
@@ -5,6 +5,11 @@
     class Client < ::Protobuf::Rpc::Connectors::Base
+      NACK_BACKOFF_INTERVALS = [0, 1, 3, 5, 10]
+
+      class NackError < RuntimeError
```
I don't believe this is needed/used anymore

film42 commented on lib/protobuf/nats/client.rb:
```diff
@@ -131,35 +152,40 @@ def nats_request_with_two_responses(subject, data, opts)
 
             # Wait for the ACK from the server
             ack_timeout = opts[:ack_timeout] || 5
-            with_timeout(ack_timeout) { ack_condition.wait(ack_timeout) }
```
I think you can delete `with_timeout` now.

film42 commented on lib/protobuf/nats/server.rb:
```diff
@@ -55,7 +55,12 @@ def enqueue_request(request_data, reply_id)
+          nats.publish(reply_id, ::Protobuf::Nats::Messages::ACK)
+        else
+          # pending client support being rolled out
+          # nats.publish(reply_id, ::Protobuf::Nats::Messages::NACK)
```
Existing clients should drop a NACK fail to continue even if a NACK is given since no ACK is received. You should be able to roll this out as a non-breaking change.

film42 commented on lib/protobuf/nats/client.rb:
```diff
@@ -29,6 +29,30 @@ def ack_timeout
       end
 
+      def nack_backoff_intervals
+        @nack_backoff_intervals ||= if ::ENV.key?("PB_NATS_CLIENT_NACK_BACKOFF_INTERVALS")
```
Adding for you ❤️

film42 commented:
Released as version `0.6.0.pre1`.
